### PR TITLE
[pinmux/padctrl] First round of (mostly CSR facing) pinmux updates

### PR DESCRIPTION
**Pull request by msfschaffner**
Note that this is a WIP PR and there are still several things I need to change.

Putting this up to enable early feedback only.

### CODE REVIEW

**Comment by msfschaffner** [comment]
Please do not review just yet - I will ping you once this is (sort of) stable.

File: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv
Comment (cindychip) [inline hw/top_earlgrey/rtl/autogen/top_earlgrey.sv:1546]: Hey Michael, I think the template has a typo here, you missed a `,`

File: hw/top_earlgrey/data/top_earlgrey.sv.tpl
Comment (msfschaffner) [reply hw/top_earlgrey/data/top_earlgrey.sv.tpl:77]: Thanks @cindychip, fixed!

File: hw/top_earlgrey/rtl/autogen/top_earlgrey.sv
Comment (msfschaffner) [reply hw/top_earlgrey/rtl/autogen/top_earlgrey.sv]: Fixed!

File: hw/top_earlgrey/rtl/top_earlgrey_asic.sv
Comment (cindychip) [inline hw/top_earlgrey/rtl/top_earlgrey_asic.sv:61]: Hey Michael, thanks for the quick update, do you also want to remove the `padctrl`(line 68) from this top_earlgrey_asic module?

Comment (msfschaffner) [reply hw/top_earlgrey/rtl/top_earlgrey_asic.sv:61]: Hey @cindy, the padring is something we still need, but I need to figure out where to put it. 
There is a chance that this will become an entirely top-dependent module that is being generated as part of the topgen flow.

File: hw/ip/pinmux/data/pinmux.hjson.tpl
Comment (tjaychen) [inline hw/ip/pinmux/data/pinmux.hjson.tpl:283]: could we change the description to something like..."for each muxable pad, select peripheral..."? It might just be me, but I always get kind of confused by just this direction, it makes me wonder whether it's pad->mio->peripheral (output from pinmux) or peripheral->mio->pad (ouput to pads)

File: hw/ip/pinmux/data/pinmux.hjson.tpl
Comment (tjaychen) [inline hw/ip/pinmux/data/pinmux.hjson.tpl:491]: do you want to add a note either here or in the spec that if an always on peripherals is connected, it should just `EN` to 0? is that the idea?

File: hw/ip/pinmux/data/pinmux.hjson.tpl
Comment (tjaychen) [inline hw/ip/pinmux/data/pinmux.hjson.tpl:676]: is there a need for a disabled mode if there's also a `wkup_detector_en`?

File: hw/ip/pinmux/rtl/pinmux.sv
Comment (tjaychen) [inline hw/ip/pinmux/rtl/pinmux.sv:203]: can you file an issue on me to figure out exactly which pwrmgr signal to connect? There isn't anything existing that fits..because the first thing pwrmgr does is shut off the clocks, which would mess up this logic unless you want to put it on the `clk_aon_i` domain.

File: hw/ip/pinmux/rtl/pinmux.sv
Comment (tjaychen) [inline hw/ip/pinmux/rtl/pinmux.sv:295]: since we're now allowed to put logic near the pad ring, why not just put it there?

File: hw/ip/pinmux/data/pinmux.hjson.tpl
Comment (msfschaffner) [reply hw/ip/pinmux/data/pinmux.hjson.tpl:283]: yeah this direction is confusing, let me change the description.

File: hw/ip/pinmux/data/pinmux.hjson.tpl
Comment (msfschaffner) [reply hw/ip/pinmux/data/pinmux.hjson.tpl:491]: yeah that's the idea. let me add that to the description.

File: hw/ip/pinmux/data/pinmux.hjson.tpl
Comment (msfschaffner) [reply hw/ip/pinmux/data/pinmux.hjson.tpl:676]: no this seems redundant. let me remove that.

File: hw/ip/pinmux/rtl/pinmux.sv
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux.sv:295]: yeah that makes sense, will move it to the padring.

File: hw/ip/pinmux/rtl/pinmux.sv
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux.sv:203]: jep, see here: #5198

**Comment by msfschaffner** [comment]
I took the compact REGWEN feature out for now in order to unblock SW.
So from my side this is good for review.

There are a couple of internal things that still need updating after this went in, but at least the CSR side should be relatively final, except for the fact that we may switch over to compact REGWENs once the regtool supports that.

CC @GregAC @silvestrst

**Comment by msfschaffner** [comment]
I am going to squash some of these commits, once approved.

**Comment by tjaychen** [comment]
michael can you double check the expectations on the pinmux initialization
hasn't changed? (i remember seeing some changes there).
if the pinmux / gpio combo messes up, the fpga never gets the boot strap
signal, and as a result it will never load.

Weicai also created a spi bootstrap test, run that one in DV and see if it
flags anything in terms of the hardware behavioral setup.

**Comment by msfschaffner** [comment]
I'll double check - however apart from the compact/noncompact multireg change, the pinmux configuration should not have to be different after this change (there are no changes to the #MIOs and pinout in this PR). I've adapted the pinmux initialization function to reflect the compact -> noncompact change.

**Comment by msfschaffner** [comment]
Good call @tjaychen!
Looks like I optimized too much away when I removed the IO_POK signals:
```systemverilog
- assign mio_data_mux = AlignedMuxSize'({(&io_pok_i) ? mio_in_i : '0, 1'b1, 1'b0});
+ assign mio_data_mux = AlignedMuxSize'(mio_in_i);
```
Find the error ;)...

**Comment by tjaychen** [comment]
ahhhh.. .the whole thing shifted down by 2? yeah that would definitely mess
things up :)

**Comment by tjaychen** [comment]
@msfschaffner if you haven't done so already, could you update the spreadsheet containing the memory map to remove `padctrl`?

**Comment by msfschaffner** [comment]
Sure will do!

If this looks good to you, I will squash most of this into one commit to make this giant thing atomic.

**Comment by msfschaffner** [comment]
The latest force push squashes some of the commits together to make them atomic.

**Comment by tjaychen** [comment]
it looks good to me.
sorry i forgot to respond earlier.

**Comment by msfschaffner** [comment]
No worries. Can you push the green button ;)?

File: hw/ip/pinmux/rtl/pinmux_wkup.sv
Comment (tjaychen) [inline hw/ip/pinmux/rtl/pinmux_wkup.sv:127]: i forgot to ask this earlier. but if someone didn't want any debounce, and were okay with this anytime this signal is high or low, would they just a really low threshold count? (0 or 1?)

**Review by tjaychen — APPROVED** [review]

File: hw/ip/pinmux/data/pinmux.hjson.tpl
Comment (msfschaffner) [inline hw/ip/pinmux/data/pinmux.hjson.tpl:727]: FYI: @tjaychen this is the register I have been talking about in the other comment

File: hw/ip/pinmux/rtl/pinmux_wkup.sv
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux_wkup.sv:127]: The pulse counter is separate from the debouncing logic.
In fact, we've got a `prim_filter` in each wakeup detector, which can be enabled or desabled via a CSR (see separate comment).

File: hw/ip/pinmux/rtl/pinmux_wkup.sv
Comment (msfschaffner) [inline hw/ip/pinmux/rtl/pinmux_wkup.sv:57]: FYI @tjaychen: see on line 75ff below

File: hw/ip/pinmux/rtl/pinmux_wkup.sv
Comment (tjaychen) [reply hw/ip/pinmux/rtl/pinmux_wkup.sv:127]: o sorry, i should correct myself then.  I guess I meant if someone wanted any length of a pulse to trigger. 
Would they basically set this to 1?

**Comment by msfschaffner** [comment]
Thanks for the review, pulling this in now!

File: hw/ip/pinmux/rtl/pinmux_wkup.sv
Comment (msfschaffner) [reply hw/ip/pinmux/rtl/pinmux_wkup.sv:127]: Yeah that's right, you could do that or just use the edge detector feature instead.

**Pull request merged**
